File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/item/ItemDescriptor.java
package top.likoslupus.cellulosesz.api.item;

import java.util.Locale;

import static top.likoslupus.cellulosesz.api.validation.NumericChecks.requirePositive;
import static top.likoslupus.cellulosesz.api.validation.TextChecks.requireNonBlank;

import static java.util.Objects.requireNonNull;

/** Registry-validated item input together with the requested business count. */
public record ItemDescriptor(
        String item,
        int count,
        String argument
) {

    public ItemDescriptor(String item, int count) {
        this(item, count, item);
    }

    public ItemDescriptor {
        item = normalizeItem(item);
        requirePositive(count, "count");
        var normalizedArgument = requireNonNull(argument, "argument").trim();
        argument = normalizedArgument.isBlank()
                ? item
                : normalizedArgument;
    }

    private static String normalizeItem(String value) {
        var normalized = requireNonBlank(
                requireNonNull(value, "item")
                        .trim()
                        .toLowerCase(Locale.ROOT), "item"
        );
        return normalized.indexOf(':') < 0
                ? "minecraft:" + normalized
                : normalized;
    }

    public String normalizedItem() {
        return item;
    }

    public String normalizedArgument() {
        return argument;
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/item/ItemDescriptor.kt
package top.likoslupus.cellulosesz.api.item

import top.likoslupus.cellulosesz.api.validation.requireNonBlank
import top.likoslupus.cellulosesz.api.validation.requirePositive
import java.util.*

/** Registry-validated item input together with the requested business count. */
@JvmRecord
public data class ItemDescriptor(
    public val item: String,
    public val count: Int,
    public val argument: String
) {

    public constructor(item: String, count: Int) : this(
        normalizeItem(item),
        count,
        normalizeItem(item)
    )

    init {
        count.requirePositive { "count" }
    }

    public fun normalizedItem(): String = item

    public fun normalizedArgument(): String = argument

    public companion object {

        @JvmStatic
        public fun normalizeItem(value: String): String {
            val normalized = value
                    .trim()
                    .lowercase(Locale.ROOT)
                    .requireNonBlank { "item" }
            return if (normalized.indexOf(':') < 0) {
                "minecraft:$normalized"
            } else {
                normalized
            }
        }

    }

}
